### PR TITLE
Add REQUIRED comment to struct fields

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
For the properties marked as required, add a REQUIRED prefix to the doc
comment.

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@jhendrixMSFT perhaps a weird thought, but is it worth grouping these properties into Required and then Optional (and then Read Only) so that it's clearer which are required/optional?

**Comment by jhendrixMSFT** [comment]
Great suggestion, I'll make this change.

File: src/generator/models.ts [hunk 1]
```diff
@@ -132,6 +132,16 @@ class StructDef {
     if (this.Language.errorType) {
       text += '\traw string\n';
     }
+    // group fields by required/optional/read-only in that order
```
Comment (catalinaperalta) [inline src/generator/models.ts:135]: Just confirming the below if is also checking in this order?

File: test/autorest/additionalpropsgroup/zz_generated_models.go [hunk 2]
```diff
@@ -96,14 +104,20 @@ type PetAPInPropertiesResponse struct {
...
 	AdditionalProperties map[string]*string
 
-	// Dictionary of
+	// OPTIONAL; Dictionary of
 	AdditionalProperties1 map[string]*float32 `json:"additionalProperties,omitempty"`
-	ID                    *int32              `json:"id,omitempty"`
-	Name                  *string             `json:"name,omitempty"`
-	OdataLocation         *string             `json:"@odata.location,omitempty"`
+
+	// Name - OPTIONAL
```
Comment (catalinaperalta) [inline test/autorest/additionalpropsgroup/zz_generated_models.go:119]: What's the difference between some having the OPTIONAL/REQUIRED/READ-ONLY at the beginning vs some after the name?  Should this be more homogenous? It looks a little weird that the Optional and required switch places in some cases.

Comment (catalinaperalta) [reply test/autorest/additionalpropsgroup/zz_generated_models.go:119]: Also, unrelated to this PR but the `Dictionary of` is missing something?

File: src/generator/models.ts (hunk 1 above)
Comment (jhendrixMSFT) [reply src/generator/models.ts:135]: Correct.  The properties are already lexicographically sorted.  This sort groups them by their required/optional/read-only status.

File: test/autorest/additionalpropsgroup/zz_generated_models.go (hunk 2 above)
Comment (jhendrixMSFT) [reply test/autorest/additionalpropsgroup/zz_generated_models.go:119]: The `Dictionary of` looks to be an issue with M4.  We don't generate this, nor is it in the swagger.  I'll follow up on that.

The modifier showing up as a suffix is for cases where the property has no doc comment.  We did it like this in track 1 as simply having `// READ-ONLY` would make the linter unhappy.  That no longer appears to be the case.  For consistency, I can remove this.

Comment (catalinaperalta) [reply test/autorest/additionalpropsgroup/zz_generated_models.go:119]: hmm, yeah at least for now, if the linter doesnt complain about it, then I vote for more consistency in the comments.

Comment (jhendrixMSFT) [reply test/autorest/additionalpropsgroup/zz_generated_models.go:119]: Also I think I misspoke about the `Dictionary of` thing, it looks like we're somehow munging the description.  I'll fix that too.

Comment (jhendrixMSFT) [reply test/autorest/additionalpropsgroup/zz_generated_models.go:119]: Fixing the `Dictionary of ` looks to be more involved than I thought.  I've opened https://github.com/Azure/autorest.go/issues/623 to track it.

Comment (jhendrixMSFT) [reply test/autorest/additionalpropsgroup/zz_generated_models.go]: The suffix thing has been fixed.

File: test/autorest/additionalpropsgroup/zz_generated_models.go [hunk 3]
```diff
@@ -176,12 +182,14 @@ type PetAPInPropertiesWithAPStringResponse struct {
 }
 
 type PetAPObject struct {
-	// Contains additional key/value pairs not defined in the schema.
+	// REQUIRED
+	ID *int32 `json:"id,omitempty"`
+
+	// OPTIONAL; Contains additional key/value pairs not defined in the schema.
```
Comment (jhendrixMSFT) [inline test/autorest/additionalpropsgroup/zz_generated_models.go:188]: I intentionally left `OPTIONAL` for additional properties as they are by definition optional.

**Review by catalinaperalta — APPROVED** [review]
Not sure if the // REQUIRED comment over a group of fields that dont have a comment over them will ever cause confusion (about whether those subsequent fields are also required), but we can get feedback on that if it becomes a problem.

**Pull request merged**